shared-unpackaged/is-obscured.js

```diff
@@ -26,55 +34,70 @@ export default function isObscured(element: Element): boolean {
-    // The bottom right corner is one less than the bounds, otherwise we
-    // can end up getting the parent of the anchor, rather than the anchor
-    // itself.
-    const anchorRight =
-        anchorRect.right -
-        parseFloat(style.marginRight) -
-        parseFloat(style.paddingRight) -
-        parseFloat(style.borderRightWidth);
-    const anchorBottom =
-        anchorRect.bottom -
-        parseFloat(style.marginBottom) -
-        parseFloat(style.paddingBottom) -
-        parseFloat(style.borderBottomWidth);
-    const bottomRightElement = document.elementFromPoint(
-        anchorRight,
-        anchorBottom,
-    );
-    // TODO(somewhatabstract): Need to cater to the case where the viewport is
-    // zoomed such that both corners are off screen but the rest isn't as in
-    // some browsers, elementFromPoint then doesn't return the element (see
-    // WB-300).
 
-    // To cope with us hitting a child of our anchor or a parent due to
-    // borders and things, we do some descendancy checks. We're ok with
-    // saying that we're visible if we hit a parent because we already
-    // checked them for visibility earlier.
-    const topLeftVisible =
-        topLeftElement &&
-        (element.contains(topLeftElement) || topLeftElement.contains(element));
-    const bottomRightVisible =
-        bottomRightElement &&
-        (element.contains(bottomRightElement) ||
-            bottomRightElement.contains(element));
-    return !topLeftVisible && !bottomRightVisible;
+    // TODO(somewhatabstract, WB-300): Need to cater to the case where the
+    // viewport is zoomed such that both corners are off screen but the rest
+    // isn't. In this case some browsers don't return the element from
+    // `elementFromPoint` then doesn't return the element.
+    // Also, consider how we might mitigate the pointer-events issue and make
+    // this call more robust.
+    const bounds = getBounds(anchorElement);
+
+    // This method does the main work, taking some coordinates and determining
+    // if our element is visible at that point or not.
+    const isVisible = (x: number, y: number): boolean => {
+        const elAtPoint = document.elementFromPoint(x, y);
+        if (
+            elAtPoint != null &&
+            (elAtPoint === popperElement || popperElement.contains(elAtPoint))
+        ) {
+            // Oh no, we're being obscured by our own popper.
+            // We need to look behind it. Shenanigans time.
+            const pointerEventsStyle = elAtPoint.style.pointerEvents;
+            // Remove pointer events so that we can look through it.
+            elAtPoint.style.pointerEvents = "none";
+            try {
+                const visible = isVisible(x, y);
+                return visible;
+            } finally {
+                // Make sure we put things back the way we found them. :)
+                elAtPoint.style.pointerEvents = pointerEventsStyle;
+            }
+        }
```

Comment:
This is the fix for the bug. It turned out that sometimes (especially when the anchor was scrolling off the top of the screen or was in the top-left region of the screen), the popper itself (or a child thereof) would be the result of `elementFromPoint` for the bottom right check. So, now if that occurs, we briefly deprive the popper (or the child) of pointer events and reperform the check, making sure to put the pointer events back to what they were when we're done.

This then means that we get a proper check of the anchor visibility based on non-anchor/popper elements, and all works as we want.